[Guok Wei Jie] iP

guokweijie opened the pull request:
# Duke Bot
>“Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

##

Duke Bot frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ **SUPER FAST** to use

##

All you need to do is,

1. download it from [here](https://github.com/guokweijie/ip/releases/tag/A-Jar).
2. double-click it.
3. add your tasks
4. let it manage your tasks for you 😉

##

Features:
- [X] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

##

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```ruby
public static void main(String[] args) {
        try {
            Duke dukeBot = new Duke(new TaskList(new ArrayList<>(100)), "taskList.txt");

            dukeBot.runUi();

            dukeBot.parse();
        } catch (IOException e) {
            System.out.println(e.getMessage());
        }
}
```

guokweijie committed:
completed Level 1 of ip

guokweijie committed:
completed Level-2

guokweijie committed:
added Task class and completed Level-3

guokweijie committed:
added Todo, Event and Deadline classes, completed Level-4

guokweijie committed:
performed automated text UI testing, completed A-TextUiTesting

guokweijie committed:
handled exceptions and added delete function, completed Level-5 and Level-6

guokweijie committed:
used try-catch block to handle exceptions, modified for Level-5

guokweijie committed:
Save tasks in taskList.txt and updates it when task list changes

guokweijie committed:
Change the Duke bot methods to non-static and add functions

guokweijie committed:
Store date and time for event and deadline

guokweijie committed:
Merge branch 'branch-Level-7'

guokweijie committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Duke.java

guokweijie committed:
Extract classes to make the code more OOP

guokweijie committed:
Put all classes in duke package

guokweijie committed:
Add gradle files to project

guokweijie committed:
Add JUnit tests for ToDo, Deadline and Event classes

guokweijie committed:
Change main class name of build.gradle file and modify loadUpData() function

guokweijie committed:
Add JavaDoc comments to all classes.

guokweijie committed:
Modify code to follow Java coding standard

guokweijie committed:
Add function to find task by searching for a keyword

guokweijie committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/DukeException.java
#	src/main/java/duke/Event.java

guokweijie committed:
Merge branch 'branch-Level-9'

guokweijie committed:
Import FileNotFoundException in Storage class

guokweijie committed:
Add on to the Javadocs to follow coding standard.

guokweijie committed:
Make Parser class set date and time instead

PeiYee88 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,211 @@
+            Scanner sc = new Scanner(System.in);
+
+            while (sc.hasNext()) {
+                try {
```
another alternatives is to use a switch case block here, but yours is totally fine.

PeiYee88 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,211 @@
+
+                            this.taskList.addTask(deadLineTask);
+
+                            System.out.println("Got it. I've added this task:\n "
```
You can have a class that handles all the header and footer of the command, so that you wont have to type it out for every single task type.

guokweijie committed:
Rebuild gradle project

PeiYee88 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,63 @@
+package duke;
```
Well done! Code for every class is clean! OOP is well-used :)

PeiYee88 commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,61 @@
+     */
+    public Deadline(String description, String by) {
+        super(description);
+        this.by = by;
```
perhaps using 'deadline' rather than 'by' is a better choice here?

PeiYee88 reviewed (commented):
All good! Just some minor edits!

Isaaclhy00 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,63 @@
+
+        this.storage.loadUpData();
+
+    }
```
Perhaps could follow the example for public Duke(String filePath) and public void run() given in week 3's A-More OOP section

Isaaclhy00 commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,61 @@
+ * A class that creates deadline task.
+ */
+public class Deadline extends Task {
+    protected String by;
```
should the variable 'by' be more descriptive of the content it holds such as 'deadlineDescription'?

Isaaclhy00 commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,61 @@
+     */
+    public void parseTime(String time) {
+        this.time = LocalTime.parse(time);
+    }
```
Perhaps parsing should be handled by the parser object as the name suggests

Isaaclhy00 commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,60 @@
+ * A class that creates the Event task.
+ */
+public class Event extends Task{
+    protected String at;
```
Should 'at' be more descriptive of its contents sucha s 'eventDescription'?

Isaaclhy00 commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,60 @@
+     *
+     * @param time Time to parse in
+     */
+    public void parseTime(String time) {
```
Perhaps parsing should be handled by the parser object as the name suggests

Isaaclhy00 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,211 @@
+/**
+ * A class that handles the user input commands.
+ */
+public class Parser {
```
Would it be better for the parser to solely handle parsing of the input? Perhaps a separate Command class can handle the execution of the commands and the UI class can handle displaying of messages.

Isaaclhy00 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,136 @@
+        this.previousTaskList = previousTaskList;
+
+        if (!Files.exists(Path.of(pathName))) {
+            file.createNewFile();
```
I like that you handled the case if the file does not already exists. However is the pathName OS-independent as required by the project outline? (Could use Paths.get() to fetch the relative file path then append the given filePath at the back to form the full file path for storage)

Isaaclhy00 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,136 @@
+     * @return List of tasks with the keyword.
+     * @throws FileNotFoundException If file does not exist.
+     */
+    public String findTasks(String keyword) throws FileNotFoundException {
```
Perhaps this shoudl be handled by the Tasklist/Task classes

Isaaclhy00 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,136 @@
+/**
+ * A class that creates the Storage object.
+ */
+public class Storage {
```
Could help if there are separate TaskList encoder and decoder classes to handle conversion of tasks to text and back.

Isaaclhy00 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,125 @@
+        this.index += 1;
+    }
+
+    /**
```
Are these methods really necessary?

Isaaclhy00 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,23 @@
+/**
+ * A class that handles the user interface of the Duke bot.
+ */
+public class Ui {
```
Maybe this UI class could handle displaying of all messages instead of just welcome and goodbye

Isaaclhy00 reviewed (commented):
Overall looks good and runs as expectdd. Weeks 3's project outline provided an Address Book example which examplifies really good OOP practices and organises all its classes and packages in a clear and concise manner. Each class and package also handles exactly what it should do and its overall structure aligns with the example codes given by the project outline, quite a good reference to check out.

guokweijie committed:
Add a GUI to Duke

guokweijie committed:
Use FXML in the application

guokweijie committed:
Merge branch 'branch-Level-10'

guokweijie committed:
Use assertions

No usage of assertions in the code.

Assertions can help to check for important assumptions that should hold
at various points in the code.

Let's use assertions in the code.

Using assertions improves the quality of code as it is a complementary
way of handling errors.

guokweijie committed:
Improve code quality for all files

The code has long methods, deep nesting and lack of abstraction.

The code needs to be more readable and follow the Java coding standards.

Let's refactor out code, make new methods and separate code into
different logical units for better readability.

This is such that the code appears more readable and adheres to coding
standards.

guokweijie committed:
Merge pull request #1 from guokweijie/branch-A-Assertions

Use assertions

guokweijie committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/duke/Duke.java
#	src/main/java/duke/Parser.java
#	taskList.txt

guokweijie committed:
Change taskList file

guokweijie committed:
Merge pull request #2 from guokweijie/branch-A-CodeQuality

Improve code quality for all files

guokweijie committed:
Attach priorities to tasks

guokweijie committed:
Merge branch 'branch-C-Priority'

guokweijie committed:
Edit JavaDoc method comments

guokweijie committed:
Implement checkstyle

guokweijie committed:
Edit comments in code

guokweijie committed:
Handle case when user do not input date and time

guokweijie committed:
Change parse method for date and time

guokweijie committed:
Change the way Event and Deadline stores date and time

guokweijie committed:
Change display pictures for Duke and user

guokweijie committed:
Add a brief User Guide

guokweijie committed:
Edit User Guide commands

guokweijie committed:
Change Ui screenshot

guokweijie committed:
Edit the User Guide

guokweijie committed:
Change the build.gradle file

guokweijie committed:
Change the JavaFX version in build.gradle

guokweijie committed:
Edit the User Guide
